### ISSUE TITLE
Restrict visibility of google/cloud/*/BUILD targets

### ISSUE THREAD

**Issue by devjgm** [googleapis/google-cloud-cpp#3701]
In https://github.com/googleapis/google-cloud-cpp/issues/3532 we're adding user-facing bazel build targets to the top-level https://github.com/googleapis/google-cloud-cpp/blob/master/BUILD file. In order to encourage/require users to only use the build targets in this top-level file we should do the following:

1. Initially, have the top-level *and* google/cloud/*/BUILD targets work for customers.
1. Message to users that this build-target change is coming, and that users should change their build targets from `//google/cloud/$x:$x_client` -> `//:$x_client`
1. Wait some period of time. 60 days? 90 days?
1. Change the default build visibility to `"//:__subpackages__"` in `google/cloud/*/BUILD`.

**Comment by coryan** [comment]
An observation: if we want to hide some targets (e.g. `//google/cloud/spanner:spanner_client_testing`) we may want to mark them as deprecated ahead of time.

**Comment by devjgm** [comment]
Ultimately we want to hide everything in `//google/...`. We want to expose _only_ the top-level `//:library` targets

**Comment by coryan** [comment]
Yeah, I am just saying that this will require the usual amount of planning.  I am trying to keep the scope of #5726 as small as possible, so I am not doing anything to hinder this bug, but I am not doing anything to make it easier either.

**Comment by coryan** [comment]
We also need to think of experimental features like GCS+gRPC: is is on a separate library target, maybe we should name it `//experimental:gcs_grpc` or something `//:experimental_gcs_grpc`.

**Comment by devjgm** [comment]
Then to

> An observation: if we want to hide some targets (e.g. //google/cloud/spanner:spanner_client_testing) we may want to mark them as deprecated ahead of time.

I guess I'd say "yes". Any target in `google/cloud/.../BUILD` should be marked deprecated early so that we can change their default build visibility later.

Actually, maybe there's nothing we need to do. Our https://github.com/googleapis/google-cloud-cpp#api-breaking-changes doc says 

> Any file or symbol that lives within a directory or namespace containing "internal", "impl", "test", "detail", "benchmark", "sample", or "example", is explicitly not part of our public API.

Would this apply to `//google/cloud/foo:foo_test`, et al? I think the spirit of that wording was that it should apply.

WDYT?

**Comment by coryan** [comment]
>> Any file or symbol that lives within a directory or namespace containing "internal", "impl", "test", "detail", "benchmark", "sample", or "example", is explicitly not part of our public API.

> Would this apply to //google/cloud/foo:foo_test, et al? I think the spirit of that wording was that it should apply.
> WDYT?

I think we do not want to play language lawyers with our customers.  In hindsight (as this bug implies) all these targets should have had restricted visibility, we made a mistake in not doing so.  Now we should mark them as deprecated, and then hide them.  Maybe faster than the usual deprecation as they were not intended as public in the first place, don't know exactly how fast, at a guess 6 months or something.

**Comment by devjgm** [comment]
Ok, then I do think _all_ existing `//google/cloud/.../BUILD` targets should be marked "deprecated" so that we can get to a restricted build visibility by default.

That's what this bug is about. You don't need to do this work as part of https://github.com/googleapis/google-cloud-cpp/issues/5726

**Comment by devjgm** [comment]
aaaaaand, that was reverted in https://github.com/googleapis/google-cloud-cpp/pull/8561.

The reason was because our builds showed tons of deprecation warnings from bazel because _we_ are still using the deprecated targets (which is intentional). But we don't want to flood our builds w/ these deprecation warnings, and we don't want our customers to see deprecation warnings due to our code using deprecated targets.

So, I think this means
* We cannot use `default_deprecation` at the file-level. We need to explicitly deprecate each target.
* We should create new private targets that are not deprecated that we use, and then create the old targets (named as they were) so customers aren't broken and mark *those* as deprecated. We should not then use the deprecated targets.

**Comment by devjgm** [comment]
OK, there are now a bunch of deprecated bazel targets that can be deleted after 2023-04-01. So we can revisit this issue then.

**Comment by coryan** [comment]
Still waiting for the passage of time.

**Comment by scotthart** [comment]
This will be available in v3.
